thermalnetwork/geometry.py

```diff
@@ -156,4 +162,125 @@ def distance_to_point(pt1, pt2):
     rotated_polygon = []
     for o_poly in polygon:
         rotated_polygon.append([rotate(pt, rot_ang, origin) for pt in o_poly])
+
+    # ensure all points are positive, shift if not
+    x_shift = 0
+    y_shift = 0
+
+    for this_polygon in rotated_polygon:
+        for coord_pair in this_polygon:
+            if coord_pair[0] < 0:
+                x_shift = max(abs(coord_pair[0]), x_shift)
+            if coord_pair[1] < 0:
+                y_shift = max(abs(coord_pair[1]), y_shift)
+
+    rotated_polygon = [[(x + x_shift, y + y_shift) for x, y in this_polygon] for this_polygon in rotated_polygon]
+
     return rotated_polygon
+
+
+def alpha_shape(points, alpha=1, only_outer=True):
+    """
+    Compute the alpha shape (concave hull) of a set of points.
+    :param points: np.array of shape (n,2) points.
+    :param alpha: alpha value.
+    :param only_outer: boolean value to specify if we keep only the outer border
+    or also inner edges.
+    :return: set of (i,j) pairs representing edges of the alpha-shape. (i,j) are
+    the indices in the points array.
+
+    Initial code taken from here: https://stackoverflow.com/a/50159452
+
+    """
```

Comment:
If we're allowing mixing/matching of network design methods and pre-sized GHE, we will run into cases where some GHE are pre-defined, and others are not. I that case, our scheme splitting the network loads based on GHE area will still need the areas of all GHE.

This code was poached from StackOverflow, and is used to find the edges of the pre-defined borehole domains, since we're allowing the users to only pass in the xy-coordinate pairs for each borehole. As such, the polygon containing each GHE could be convex or non-convex. This algorithm was needed to find the polygon edges and uses ScyPy's `Delaunay` tessellation. 

@chriswmackey let me know if you have a better suggestion for this.

Reply:
@chriswmackey We are not actually using the default `alpha` value. See the link here where we are brute force searching for the alpha that gives the most edge vertices.

https://github.com/NREL/ThermalNetwork/pull/63/files#diff-6391a9a0ccbe3dea8eaccbc2e869d544dd6ef08c68375e6d18379a7de1f351b9R271-R274

I agree that it's not a perfect solution for a variety of reasons, but I think we just accept it for now and move on. Let's keep this in mind, and hopefully find non-scipy solution soon. Note, that I have only tested it with the L-shaped field in the test, so perhaps this will scale poorly once we go to much larger pre-designed fields.